Add MiMa checks and fix binary incompatible changes

kyri-petrou opened the pull request:
This PR adds MiMa checks during CI. See comments below for some more info. Just some points for discussion:

1. Should we be strict when the code is source compatible but not binary compatible? (e.g., the `makeScalar` method wasn't bincompat)
2. What's the best way to ensure that we clear the MiMa filters after each release?

kyri-petrou committed:
Overload `scalarSchema` method to revert breaking change

kyri-petrou committed:
Fix compiling error

kyri-petrou committed:
Add mima checks and fix bincompat issues

kyri-petrou commented on .circleci/config.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
       - restore_cache:
           key: sbtcache
-      - run: sbt ++3.2.2 core/test catsInterop/compile benchmarks/compile monixInterop/compile clientJVM/test clientJS/compile zioHttp/test tapirInterop/test http4s/test federation/test reporting/test tracing/test
+      - run: sbt ++3.2.2 core/test catsInterop/compile benchmarks/compile monixInterop/compile clientJVM/test clientJS/compile zioHttp/test tapirInterop/test http4s/test federation/test tools/test reporting/test tracing/test
```
We were only testing `tools` against Scala 2.12. Was that intentional?

ghostdogpr replied:
Probably an oversight 👍

kyri-petrou commented on core/src/main/scala/caliban/schema/Types.scala:
```diff
@@ -7,14 +7,19 @@ import scala.annotation.tailrec
+  ): __Type =
+    makeScalar(name, description, specifiedBy, None)
+
+  def makeScalar(
+    name: String,
+    description: Option[String],
+    specifiedBy: Option[String],
+    directives: Option[List[Directive]]
```
So this change was source compatible but not binary compatible. I'm 50-50 whether I should revert this change and add this method as a filter instead

ghostdogpr replied:
Considering the changes I'm working on with the adapters, I think we may go 2.2 directly so maybe it's fine to keep those breaking changes.

kyri-petrou replied:
> How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

@ghostdogpr I  added a flag in `build.sbt` to disable CI breaking on binary incompatible changes. The good thing about this flag is that the plugin will still report the bincompat issues in the console, so we can always look at it when we're close to releasing to know what bin / source incompatible changes we introduced

Also, Since the next release will likely be 2.2.0, I reverted the changes to fix the bincompat issue in `makeScalar`.

kyri-petrou commented on build.sbt:
```diff
@@ -579,6 +595,22 @@ lazy val commonSettings = Def.settings(
+lazy val enableMimaSettingsJVM =
+  Def.settings(
+    mimaPreviousArtifacts := previousStableVersion.value.map(organization.value %% moduleName.value % _).toSet,
+    mimaBinaryIssueFilters ++= Seq( // Filters should be cleared after a new release
+      ProblemFilters.exclude[MissingTypesProblem]("caliban.schema.SchemaDerivation$Auto"),
+      ProblemFilters.exclude[ReversedMissingMethodProblem](
+        "caliban.schema.SchemaDerivation#Auto.caliban$schema$SchemaDerivation$Auto$$$outer"
+      )
```
The changes to `Schema.Auto` are not bincompat but that's okay since it wasn't working as intended previously

ghostdogpr commented:
> 1. Should we be strict when the code is source compatible but not binary compatible? (e.g., the `makeScalar` method wasn't bincompat)

I think we should be strict for patch versions.

> 2. What's the best way to ensure that we clear the MiMa filters after each release?

I have a step where I update the version in docs, I can do it at the same time (it's manual though).

How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

kyri-petrou committed:
PR comments

kyri-petrou commented:
> How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

@ghostdogpr I  added a flag in `build.sbt` to disable CI breaking on binary incompatible changes. The good thing about this flag is that the plugin will still report the bincompat issues in the console, so we can always look at it when we're close to releasing to know what bin / source incompatible changes we introduced

Also, Since the next release will likely be 2.2.0, I reverted the changes to fix the bincompat issue in `makeScalar`.

ghostdogpr commented on core/src/main/scala/caliban/introspection/adt/__Directive.scala:
```diff
@@ -17,4 +17,7 @@ case class __Directive(
     )
+
+  @deprecated("Use isRepeatable")
+  def repeatable: Boolean = isRepeatable
```
Can we delete these then?

kyri-petrou replied:
Done 👍 My bad, had to head out in a hurry earlier and didn't cleanup properly

ghostdogpr commented on core/src/main/scala/caliban/schema/Schema.scala:
```diff
@@ -147,6 +147,15 @@ trait GenericSchema[R] extends SchemaDerivation[R] with TemporalSchema {
       override def resolve(value: A): Step[Any]                              = PureStep(makeResponse(value))
     }
 
+  @deprecated("use the overloaded method with directives instead")
```
Delete?

ghostdogpr commented on core/src/main/scala/caliban/schema/Types.scala:
```diff
@@ -7,9 +7,6 @@ import scala.annotation.tailrec
 
 object Types {
 
-  /**
```
Any reason to remove this?

kyri-petrou replied:
I deleted the extra method but didn't realise I also deleted the scaladoc 😅

kyri-petrou committed:
Revert original PR changes

kyri-petrou committed:
Merge branch 'series/2.x' into fix-breaking-change

# Conflicts:
#	project/plugins.sbt

kyri-petrou committed:
Bump resource_class for jdk213 and jdk3 tests
